File: src/hardware/tests/test_linkedtarget.py
#  Approved for public release: distribution is unlimited. PA Case Number AFRL-2023-4617.

#  © 2023 Raytheon BBN Technologies Corp. All rights reserved. Sponsored by the Air Force Research Laboratory (AFRL)  



import unittest
from mock import Mock
from uuid import getnode as get_mac_address
import paho.mqtt.client as mqtt
import time
import json
import sys
sys.path.append('..')
from target.LinkedTarget import LinkedTarget


class TestLinkedTarget(unittest.TestCase):
    # Create a new mock object
    mock = Mock()
    
    target_identifier = 'TARGET'
    

    def test_update_valid_capture_state(self):
        """
        Test whether the linked target correctly updates its capture state to captured
        """
        # Create a new target
        target = LinkedTarget(self.mock, self.mock, self.target_identifier)

        # Set the payloads in range to be larger than the required number of payloads
        payloads_in_range = 5

        # Update the target state
        target._set_payloads_in_range(payloads_in_range)

        # Update the target capture state
        target._update_individual_capture_state()

        self.assertTrue(target._get_individual_capture_state())

        return
        

    def test_update_invalid_capture_state(self):
        """
        Test whether the linked target correctly updates its capture state to not captured
        """
        # Create a new target
        target = LinkedTarget(self.mock, self.mock, self.target_identifier)

        # Set the payloads in range to be less than the required number of payloads
        payloads_in_range = 0

        # Update the target state
        target._set_payloads_in_range(payloads_in_range)

        # Update the target capture state
        target._update_individual_capture_state()

        self.assertFalse(target._get_individual_capture_state())

        return
        

if __name__ == 'main':
    unittest.main()

File: src/analytics_server/src/test/java/com/bbn/mace/TestLinkedTarget.java
//  Approved for public release: distribution is unlimited. PA Case Number AFRL-2023-4617.

//  © 2023 Raytheon BBN Technologies Corp. All rights reserved. Sponsored by the Air Force Research Laboratory (AFRL)  



package com.bbn.mace;

import org.eclipse.paho.client.mqttv3.MqttException;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.bbn.mace.target.LinkedTarget;
import com.bbn.mace.utils.GpsLocation;
import com.google.gson.JsonArray;

import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;


public class TestLinkedTarget {
	private static LinkedTarget target;
	private static final String uid = "TARGET-123456";
	private static final String host = "tcp://test.mosquitto.org:1883";
	
	@BeforeAll
	static void setUp() throws InterruptedException {
		try {
			target = new LinkedTarget(uid, host);
		} catch (MqttException e) {
			System.out.println("An error occurred while attempting to generate a new target: ");
			e.printStackTrace();
		}
	}
	
	
	@Test
	@DisplayName("Setting Linked Target State Test")
	void testSetTargetState() throws InterruptedException {
		// Test location
		double latitude = 65.4;
		double longitude = 55.4;
		double altitude = 10.56;
		
		// Create a new test location
		GpsLocation testLocation = new GpsLocation(latitude, longitude, altitude);
		
		// Test payloads
		int payloadsInRange = 4;
		
		// Test captured flag
		boolean captured = false;
		
		// Set the state
		target.setTargetState(testLocation, payloadsInRange, captured);

		// Assert that the target properties were set properly
		assertEquals(target.getLocation().getLatitude(), testLocation.getLatitude());
		assertEquals(target.getLocation().getLongitude(), testLocation.getLongitude());
		assertEquals(target.getLocation().getAltitude(), testLocation.getAltitude());
		assertEquals(target.getPayloadsInRange(), payloadsInRange);
		assertEquals(target.getIndividualCaptured(), captured);
	}
	
	
	@Test
	@DisplayName("Setting Mass Target Configurations Test")
	void testSetTargetConfigsWithCapture() throws InterruptedException {
		// Test configuration values
		double range = 1.5;
		boolean setCaptured = true;
		boolean captured = true; // this should be set to true if working properly
		JsonArray targets = new JsonArray();
		
		targets.add(target.getUid());
		
		// Set the test configurations
		target.setTargetConfigs(range, setCaptured, captured, targets, "test-network");
		
		// Assert that the configurations were set properly
		assertEquals(target.getAcceptableDetectionRange(), range);
		assertEquals(target.getNetworkCaptured(), captured);
		assertEquals(target.getNetworkTargets(), targets);
	}	
}
